lib/manageiq/appliance_console/saml_authentication.rb

```diff
@@ -0,0 +1,212 @@
+            end
+          end
+        end
+      end
+
+      def fetch_idp_metadata
+        idp_metadata = options[:saml_idp_metadata]
+        if path_is_file(idp_metadata) && idp_metadata != IDP_METADATA_FILE
+          debug_msg("Copying IDP metadata file #{idp_metadata} to #{IDP_METADATA_FILE} ...")
+          FileUtils.cp(idp_metadata, IDP_METADATA_FILE)
+        elsif path_is_url(idp_metadata)
+          debug_msg("Downloading IDP metadata file from #{idp_metadata}")
+          download_network_file(idp_metadata, IDP_METADATA_FILE)
+        end
+      end
+
+      def copy_apache_saml_configfiles
+        debug_msg("Copying Apache SAML Config files ...")
+        copy_template(HTTPD_CONFIG_DIRECTORY, "manageiq-remote-user.conf")
+        copy_template(HTTPD_CONFIG_DIRECTORY, "manageiq-external-auth-saml.conf")
+      end
+
+      def remove_apache_saml_configfiles
+        debug_msg("Removing Apache SAML Config files ...")
+        remove_file(HTTPD_CONFIG_DIRECTORY, "manageiq-remote-user.conf")
+        remove_file(HTTPD_CONFIG_DIRECTORY, "manageiq-external-auth-saml.conf")
+      end
+
+      def configured?
+        path_join(HTTPD_CONFIG_DIRECTORY, "manageiq-external-auth-saml.conf").exist?
+      end
+
+      def restart_httpd
+        httpd_service = LinuxAdmin::Service.new("httpd")
+        if httpd_service.running?
+          say("Restarting httpd ...")
+          httpd_service.restart
+        end
+      end
+
+      # SAML IDP Metadata
+
+      def validate_saml_idp_metadata_option
+        idp_metadata = options[:saml_idp_metadata]
+        raise "Must specify the SAML IDP metadata file or URL via --saml-idp-metadata" if idp_metadata.blank?
+
+        raise "Missing SAML IDP metadata file #{idp_metadata}" if path_is_file(idp_metadata) && !File.exist?(idp_metadata)
+      end
+
+      def path_is_file(path)
+        path.present? && !path_is_url(path)
+      end
+
+      def path_is_url(path)
+        path =~ /\A#{URI.regexp(["http", "https"])}\z/
+      end
+
+      # File Management
+
+      def remove_file(*args)
+        path = path_join(*args)
+        if path.exist?
+          debug_msg("Removing #{path} ...")
+          File.delete(path)
+        end
+      end
+
+      def copy_template(dir, file)
+        src_path = path_join(template_directory, dir, file)
+        dest_path = path_join(dir, file)
+        debug_msg("Copying template #{src_path} to #{dest_path} ...")
+        FileUtils.cp(src_path, dest_path)
+      end
+
+      def download_network_file(source_file_url, target_file)
+        require "net/http"
+
+        say("Downloading #{source_file_url} ...")
+        result = Net::HTTP.get_response(URI(source_file_url))
+        raise "Failed to download file from #{source_file_url}" unless result.kind_of?(Net::HTTPSuccess)
+
+        File.write(target_file, result.body)
+      end
+
+      def template_directory
+        @template_directory ||= Pathname.new(ENV.fetch("APPLIANCE_TEMPLATE_DIRECTORY"))
```

Comment:
fetch will blow up if the ENV var is not set...is that expected?

Reply:
yep, that's fine. and it's failing early enough that no change is made to the appliance.

```
Configuring SAML Authentication for https://aab-saml-config.aabsaml.redhat.com ...
Failed to Configure SAML Authentication - key not found: "APPLIANCE_TEMPLATE_DIRECTORY"
```